### PR TITLE
Nerf overheat behavior

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature:** This PR implements the feature request detailed and discussed in issue #5976.

## Feature Details
This PR changes the mechanics of overheating to be less immediately punishing, therefore nerfing heat damage.
* Currently an overheated ship is essentially a disabled ship. In either case the ship is incapable of acting.
* With this change, an overheated ship shuts down all energy and heat generation, and causes solar collection to be cut in half. This means that an overheated ship is able to retaliate or run, and only becomes incapacitated once it is out of energy. Using solar collection allows for some counterplay to being overheated.
* An overheated ship is still capable of taking all other actions normally so long as it still has energy for them.
  * This means that cooling runs while a ship is overheated, whereas previously it did not, making overheated ships return to normal heat levels quicker than before (assuming the cooling outpaces any continued heat generated by other means).
  * This also means that active cooling will now work at its advertised level while the ship is overheated, as opposed to currently where active cooling effectively only reaches 99% of its advertised level before the ship overheats.
* Added an IsParalyzed function that checks if the ship is overheated and out of energy. This replaces various calls to IsOverheated that were made in the AI class.
* No further changes have been made to the AI or to any heat damage weapons.
  * Heat damage weapons may need to be rebalanced according to how strong heat damage now is.
  * The AI may be improved in the future as to make overheated ships (personality dependent) stop trying to fight and instead try to run, saving energy.
  * The AI may be changed as to have most ships (personality dependent) ignore overheated but not yet paralyzed ships if the above change is made.

This change alone isn't too massive a nerf to heat damage since an overheated ship that runs itself out of energy with this change is essentially the same as a overheated ship in the game currently, but I believe that this is a step in the right direction.

Opening as a draft as there may be places that we want to put IsParalyzed to use or not use it where it currently is.

## Testing Done
Attacked a Korath Raider with a bunch of flamethrowers while having a tactical scanner equipped. Observed how once the ship's heat reached over 100%, the ship was still able to fight but quickly began losing energy.

## Performance Impact
N/A

### CODE REVIEW

**Comment by Amazinite** [comment]
Should overheated ships also not be able to repair their shields or hull? While this would make damaging the ship easier, it would also mean the ship isn't spending as much energy or generating as much heat while overheated.

**Comment by Arachi-Lover** [comment]
> Should overheated ships also not be able to repair their shields or hull? While this would make damaging the ship easier, it would also mean the ship isn't spending as much energy or generating as much heat while overheated.

I think that kind of behavior would fit more with `discharge` and `corrosion` damages than with heat.

**Comment by ravenshining** [comment]
In the story, flamethrowers were intended to be used against smaller ships, but with this mechanic, a battery-only fighter would effectively be immune to heat damage, so I think *something* other than energy generation ought to be affected.

Shield generation seems like a good candidate to me - one might think of shields as locking in heat, plus it is more likely to be used by the human ships intended to be affected.  But, perhaps not hull repair.

Interfering with weapons fire would have broader applicability.  I don't think it makes as much sense as interfering with shield generation, but I think it would be best for gameplay: forcing *any* ship into a defensive stance where they can only run and heal.  If this route is taken, however, we would need to have the AI try to flee when overheated.  Also, if possible, exempting anti-missile fire might be a good idea.

On an unrelated note, not that we currently use it, but I would think `fuel energy` should also be disabled by heat.

**Comment by Amazinite** [comment]
> In the story, flamethrowers were intended to be used against smaller ships

It's not said that it's intended for flamethrowers to be used against smaller ships, just that they're more effective against smaller ships. 

> with this mechanic, a battery-only fighter would effectively be immune to heat damage, so I think something other than energy generation ought to be affected.

Battery powered flight is already a bit of a risky move, and will only really be made viable for fighters in the future with changes from Sam. Even assuming that those changes are in place, I don't see an issue with going battery powered as being a hard counter to heat damage given what you give up by going battery powered (i.e. combat up-time), particularly with respect to fighters since they're already rather weak ships as is and being able to effectively make your fighters immune to heat damage doesn't seem like a bad thing.

> Shield generation seems like a good candidate to me

I did ask if overheated ships should also disable shield generation and hull repair, so I'm open to it. But I don't think this would actually help with the battery powered fighter situation, since fighters don't often have their own shield generation, or at least not anything strong enough to make a difference; you're probably better off stacking more batteries than adding a shield generator to most fighters.

> Interfering with weapons fire would have broader applicability.

That's already what ion damage does. I don't want to have major redundancies between special damage types like that.

> I would think `fuel energy` should also be disabled by heat.

It already is under this PR's current implementation.

**Comment by Hurleveur** [comment]
As I said in the discord, I wish that you shields would protect you from heat, until they overheat/get disabled. This would mean that not completely overheating an enemy ship still does something - taking out its shields.

We could even have different kind of resistance to heat for the shields (for later PR), and my idea would be once you reach a given % of your max heat taken whilst the shields were on, they turn off and you start actually taking heat damage. This would mean an initial protection against heat, which makes sense - shields block enemy shots.
The shields would then cool down, and eventually come back online.

When the shields are down, and you take enough heat damage then you can overheat, in which case what you've done here seems good, with the addition of removing any hull repair (because otherwise you'll run out of power too quickly, you need all of it for the engines).

**Comment by Amazinite** [comment]
As with how I don't think heat damage should affect weapons as that's ion damage's area, I don't think heat damage should take out shields without needing shield damage as that's disruption's area to allow you to attack a ship's hull without going through its shields the normal way.

**Comment by Hurleveur** [comment]
It doesn't take out enemy shields it just means part of your attacks get trough, so it isn't the same - also it's barely used.

But anyway, in that case, about the hull generation and shield generation, I think they should be cut down to save energy & heat but maybe that could be an option for the player, like `sparing when overheated` or smth, that if on means everything we can disable to save on heat & energy will be. It should even turn auto-fire down when overheated, because otherwise you'll be out of power before you can even notice you did overheat.

**Comment by Hurleveur** [comment]
Also we should definitely make the AI go into sparing mode once they get overheated, and not disable themselves in a matter of seconds.
What I mean by that is only shoot ammo weapons as they produce very little heat & consume very little energy, as well as maybe try to escape the fight, with engines or by jumping. They would do the first one if they feel like they can win, and hopefully they'll stop being overheated with this sparing behavior, and behave normally again.

**Comment by Zitchas** [comment]
I definitely agree that heat should *not*be fully effective against a ship with shields up. It is a fairly simple form of energy transfer, and one that shields are presumably optimized against as it is a universally occurring hazard for ships.

**Comment by warp-core** [comment]
> I definitely agree that heat should _not_be fully effective against a ship with shields up. It is a fairly simple form of energy transfer, and one that shields are presumably optimized against as it is a universally occurring hazard for ships.

Shields already halve incoming heat damage. Are you suggesting reducing that further? Should shields block all heat damage? No heat damage through shields at all?

**Comment by Amazinite** [comment]
> It doesn't take out enemy shields it just means part of your attacks get trough, so it isn't the same

The *identity* of the two is what would be the similar. The identity of disruption damage is that it allows you to deal hull damage without needing to cut through the shields first; that is, you don't need to bother with shield damage. If a certain amount of heat damage completely destroyed an enemy's shields, the identity of that would be that heat damage allows you to deal hull damage without needing to cut through the shields normally; that is, you don't need to bother with shield damage. In either case, the importance of shield damage is diminished.

Remember also though that the changes being made here are to overheating. A ship doesn't only overheat via heat damage; ships can also overheat themselves. This means that making overheating break a ship's shields would make fighting enemies like the Korath much easier since they can be overheated often, which would cause them to break their own shields.

**Comment by Amazinite** [comment]
If we want to add more to overheating beyond just powering down the generators, which wouldn't affect battery powered ships, then we could also have overheated ships have the same behavior that low-crew ships have, which is a possibility I mentioned in #5976. This, unlike also disabling shield generators and hull repair, would be something that impacts all ships regardless of how they are fitted.

**Comment by EjoThims** [comment]
> If we want to add more to overheating beyond just powering down the power generators, which wouldn't affect battery powered ships, then we could also have overheated ships have the same behavior that low-crew ships have

I would much rather triggering low crew behavior be a separate, new damage type.

I think heat cutting gen output is probably enough, but if we want it to also directly stop shield/hull regen, maybe tie that to hitting the 200% threshold? So hot ships don't produce energy, and very, very hot ships also lose specific ways to use energy?

And I don't have a problem with certain fighter builds now being essentially immune to heat build up, because those same builds are especially weak to energy/ion damage instead.

**Comment by EjoThims** [comment]
@Amazinite Thinking about the stream yesterday, I've thought of a slightly different approach which I think may serve to reduce heat's overall power, the very binary nature of it's power, and the effectiveness against small and large craft all in one go.

Instead of making heat turn off power generation in total, have it scale down power generation over 100% (probably something akin to inverse of how active cooling scales up) so that there is 0% generation at 200% heat.

This is significantly weaker than even the current idea, though, so we also add damage. Not direct damage from heat, even when overheated (let that behavior stay special for Ka'het and other specific content later), but instead as a scaling increase to damage taken based on heat percent. Essentially negative hull/shield protection as your ship gets hotter and hotter. I'm thinking something along the lines of .005 heat percentage from 0-100% heat, and .02 heat percentage from 100-200%. But specific numbers are, of course, also subject to change. :)

With the two together, you get up to 5% extra damage before energy generation begins to scale down at 100% heat, and at 200% heat you'd have 20% extra damage caused and 0% power generation. So building and holding low amounts of heat is still generally a non-issue, building and holding high amounts of heat is something you want to for sure avoid, and as soon as you are "overheated," the bad news ramps up much faster and in a secondary way as well.

This takes heat's primary effect into a direct combat role and makes even non debilitating levels of heat cause some kind of influence in the fight, just like recent changes to ion.

It also makes heat not only still effective but especially dangerous against small craft (even battery only builds), because they require less total heat damage to ramp up the damage you're causing them, while still providing an even greater direct penalty to big ships that are able to just keep fighting from energy reserves for extended periods while so overheated as to produce no energy at all.

**Comment by Amazinite** [comment]
I do like the idea of potentially turning heat into a damage buff of sorts, but I think that's something to keep in our pocket for later rather than do that right now (or perhaps even make it a separate damage type from heat; "weakening"?). As I've mentioned, we can continue to change up heat damage later if this change isn't sufficient, as I'd rather not make too many changes at once.

File: data/coalition/coalition outfits.txt [hunk 1]
```diff
@@ -19,6 +19,7 @@ outfit "Small Collector Module"
 	"solar collection" .8
 	"solar heat" .4
 	description "Solar power is the mainstay of the Coalition's merchant marine. The design of these panels has been refined over the millennia, growing ever more efficient."
+	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overhetead, solar panels continue to work at a reduced efficiency."
```
Comment (Amazinite) [inline data/coalition/coalition outfits.txt:22]:
```suggestion
	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overheated, solar panels continue to work at a reduced efficiency."
```

File: data/coalition/coalition outfits.txt [hunk 2]
```diff
@@ -31,6 +32,7 @@ outfit "Large Collector Module"
 	"solar collection" 3.3
 	"solar heat" 1.5
 	description "One of these solar panels is enough to meet the energy needs of a medium-sized Coalition civilian ship, especially given the energy efficiency of their engines and the fact that under their laws only Heliarch ships are allowed to have weapons."
+	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overhetead, solar panels continue to work at a reduced efficiency."
```
Comment (Amazinite) [inline data/coalition/coalition outfits.txt:35]:
```suggestion
	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overheated, solar panels continue to work at a reduced efficiency."
```

File: data/human/power.txt [hunk 3]
```diff
@@ -71,6 +71,7 @@ outfit "KP-6 Photovoltaic Panel"
 	"outfit space" -2
 	"solar collection" .1
 	description "This early commercial photovoltaic system was the driving force behind humanity's first interplanetary trade network. Rugged and modular, it can be bolted to the hull of any ship and has seen even the most remote parts of human space."
+	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overhetead, solar panels continue to work at a reduced efficiency."
```
Comment (Amazinite) [inline data/human/power.txt:74]:
```suggestion
	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overheated, solar panels continue to work at a reduced efficiency."
```

File: data/human/power.txt [hunk 4]
```diff
@@ -80,6 +81,7 @@ outfit "KP-6 Photovoltaic Array"
 	"outfit space" -16
 	"solar collection" 1.1
 	description "The compact modularity of the KP-6 lent itself well to the creation of larger solar collection arrays. This original Korunak Photovoltaics model is equipped with a solar tracker, allowing it to make better use of available light."
+	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overhetead, solar panels continue to work at a reduced efficiency."
```
Comment (Amazinite) [inline data/human/power.txt:84]:
```suggestion
	description "	Solar panels are more resilient to being overheated than normal energy generation. While a normal reactor will shut down when overheated, solar panels continue to work at a reduced efficiency."
```

**Review by Amazinite — COMMENTED** [review]
Oops

**Comment by Zitchas** [comment]
Really? A fairly significant impact across a wide swath of the game, including code and mechanics, and you just leap from "draft" to "merged" in a few minutes with zero reviews or discussion of the actual "this is what I'm proposing we change" PR?

**Comment by Amazinite** [comment]
This was only a draft because I still wanted to update the references to overheating in mission dialog and outfit descriptions. The core of this PR, the changes to how overheating behaves, were discussed at length on a stream to positive feedback.

**Comment by Zitchas** [comment]
Stream positive impressions does not equal approving the code.

For myself, I assumed this was just another work in progress and was waiting for it to reach a state ready to be reviewed before taking a hard look at the changes. I don't have time to read every tweak on a draft that may sit in draft for years and change radically over that time period.

**Comment by EjoThims** [comment]
Well, to be fair, Korath stuff really did depend on some form of this change coming before the next release. Xapleaux stuff too, while not being nearly ready yet, suffers greatly from allowing heat damage to remain as effectively a soft lock.

But I still think this is too binary an implementation and hope that further review and change is indeed still coming so that we can settle on something more in line with the current implementation of ion damage.

**Comment by Hurleveur** [comment]
Given overheating was OP for so long, I think this is a change in the right direction.
However, I do agree it was done too quickly still, and it should have been clear this was a draft for no more than a few descriptions.

That said, we can further change it if there is a need, with some actual playing feedback.
Also to be fair the code changes are minimal, the real consequence is in gameplay.

**Comment by ravenshining** [comment]
> Well, to be fair, Korath stuff really did depend on some form of this change coming before the next release. Xapleaux stuff too (while not being nearly ready yet) suffers greatly from allowing heat damage to remain as effectively a soft lock.

I wouldn't say Korath stuff "depended" on this at all.  Certain people may have gotten overly worried about the heat weapons, but I think that was more worrying about what other people would think, exacerbated by some faulty assumptions, than actual worrying about the mechanics themselves.  Certainly not a justification for speed-merging without review, to say nothing of the Xapeleax which are nowhere near release ready from what I understand.

**Comment by leklachu** [comment]
Is this the right place to continue discussing the flamethrower change?

I did a little testing. I'm not the right person for serious and skilful balance questions. But I do love roaming round with my too-many-flamethrowers for casual play. They allow me to temporarily disable a baddie from attacking me or another ship, without actually taking out that enemy ship. I appreciate 1v1 duelling power of flamethrowers is over-powered, but do you guys feel the same in a bigger fight? Swapping weapons out for flamethrowers loses me damage output, which is tricky if I don't have time to drag out a 1v1 fight.

So for casual, role-play, and feeling like I'm a valuable contributor in a big FW-Navy fight, I enjoy the overpowered flamethrowers. Anyway that may be just me, that's okay. As the player says to Edwards, "It's a terrifying weapon" - terrifying, but with niche use.

But I tried out the patch now. Two flamethrowers on my Gull, fighting a sparrow. I can overheat him easily enough, but it seems to do nothing, because unless I keep focus on him forever - by which time I might as well kill him, if that's my ultimate aim - he has so much battery compared to power needs he just keeps going, keeps pace with the next-door sparrow that I didn't overheat. I tried against a few ships, upping to my 7-flamethrower model with more durability, and against most it seemed a very long time of staying overheated until their batteries ran out and it had any effect.

Perhaps the new strategy with flamethrowers has to be twinning them with ion cannons?

In principle I really like the idea of overheating stopping power source heat generation, but practically I'm not convinced. Perhaps as has been suggested, something similar to the Ion's weapon jamming: something that has some effect before the full paralysis, is good? I'm afraid I don't have any more ideas than what's already been suggested, though. And I accept you may feel it's more important to balance things for a different style of play to mine.

**Pull request merged**